test/domainPurchaser.spec.ts

```
@@ -0,0 +1,253 @@
+      registrar = await smock.smock.fake(Registrar__factory);
+      erc20Token = await smock.smock.fake(ERC20Upgradeable__factory);
+
+      const factory = new DomainPurchaser__factory(creator);
+      purchaser = await factory.deploy();
+      await purchaser.initialize(
+        erc20Token.address,
+        hub.address,
+        creator.address,
+        pricingData,
+        0
+      );
+    });
+
+    it("allows a user to purchase a network", async () => {
+      hub.getRegistrarForDomain.returns(registrar.address);
+      registrar.registerSubdomainContract.returns(1);
+
+      erc20Token.transferFrom.returns(true);
+
+      const tx = await purchaser
+        .connect(user1)
+        .purchaseSubdomain(0, "meowmeowmeow", "ipfs://Qm1");
+
+      expect(tx)
+        .to.emit(purchaser, "NetworkPurchased")
+        .withArgs(1, user1.address);
+
+      expect(registrar.registerSubdomainContract).to.have.been.calledWith(
+        0,
+        "meowmeowmeow",
+        user1.address,
+        "ipfs://Qm1",
+        0,
+        false,
+        user1.address
+      );
+    });
+
+    it("fails if the user doesn't have enough funds", async () => {
+      hub.getRegistrarForDomain.returns(registrar.address);
+      registrar.registerSubdomainContract.returns(1);
+
+      erc20Token.transferFrom.reset();
+      erc20Token.transferFrom.reverts();
+
+      const tx = purchaser
+        .connect(user1)
+        .purchaseSubdomain(0, "👻👻👻", "ipfs://Qm1", { gasLimit: 3000000 });
+
+      await expect(tx).to.be.revertedWith("SafeERC20: low-level call failed");
+    });
+
+    it("charges the user the right amount for short length network", async () => {
+      hub.getRegistrarForDomain.returns(registrar.address);
+      registrar.registerSubdomainContract.returns(1);
+
+      erc20Token.transferFrom.reset();
+      erc20Token.transferFrom.returns(true);
+
+      await purchaser
+        .connect(user1)
+        .purchaseSubdomain(0, "👻👻👻", "ipfs://Qm1");
```

Setting the subdomain name as spooky ghosties got me thinking. Do we need logic to disallow certain characters? At the very least it seems including the domain delimiter character `.` should throw an error